### DISCUSSION TITLE
Understanding filter predicates and a possible race condition?

### DISCUSSION THREAD

**Discussion by sjbodzo** [kube-rs/kube#1534] [Q&A] [answered]
When writing a fairly standard controller that watches a stream of `K: CustomResource`, I notice that the following filter does not do what I expect: 
```
    let sink_stream = reflector(sink_writer, watcher(sinks, cfg.clone()))
        .default_backoff()
        .touched_objects()
        .predicate_filter((predicates::generation).fallback(predicates::resource_version));
    Controller::for_stream(sink_stream, sink_reader)
    // start the stream and visit each item, doing things
``` 

In this case my custom resource is type `Sink`, and the `Api<Sink>` details are bog-standard / default. I **expect** that creating a `Sink` in the cluster will trigger the stream and drive the reconciliation process on this controller. 

What I **observe** is that the `predicate_filter` completely misses the new object creation. The same holds true if I remove the `fallback` from the `predicate_filter` as well. If I remove the `predicate_filter`, then the stream triggers on `Sink` creation as I would expect. 

Some open questions: 
- Could this be a race where the object triggers before the initial metadata is populated in the object? The Kubernetes SIG spec uses "should" as a weak constraint on the logical clock (e.g. `generation`, `resourceVersion`) fields, so perhaps they are not set before the object is filtered out of the queue?
- If I delete the pod running my controller, the new pod that comes up _immediately_ triggers on the `Sink` creation event and runs the reconciliation loop. Does this give credence to the race condition theory? 
- If I remove the `predicate_filter` I do not observe infinite re-triggering, which I would expect given that the controller watches `Sink` resources and updates the status subresource of the `Sink` in its reconciliation loop. I find this result confusing, given that nothing in `touched_objects` suggests this behavior. Why doesn't this cause re-triggering on the same stream? Perhaps `Event::Applied<K>` does not include updates to the subresource? 

Thanks in advance for your time / help!

**Comment by clux** [comment]
some small answers, i don't know exactly what's going on here.

> Could this be a race where the object triggers before the initial metadata is populated in the object?

if that's the case, then that would be a pretty big footgun in predicates. i've never seen objects missing data population like this, but I've not particularly looked for it either.

if you have an environment to test, a `.map()` with a `println("saw unfiltered {obj:?}")` after the predicate filter should help confirm / deny this.

> If I delete the pod running my controller, the new pod that comes up immediately triggers on the Sink creation event and runs the reconciliation loop.

This part makes sense. The new controller will not have anything built up in its [temporary memory predicate cache](https://github.com/kube-rs/kube/blob/f9902f1439b3c0baafc2ece1680644c2bfade742/kube-runtime/src/utils/predicate.rs#L106) so every new event flows through unfiltered once.

> If I remove the predicate_filter I do not observe infinite re-triggering, which I would expect given that the controller watches Sink resources and updates the status subresource of the Sink in its reconciliation loop. I find this result confusing, given that nothing in touched_objects suggests this behavior. Why doesn't this cause re-triggering on the same stream? Perhaps Event::Applied<K> does not include updates to the subresource?

Without a filter, that would indeed be weird **if** your reconciler is writing __non deterministic updates__ to the status object - as that should indeed re-trigger.  It's a normal `watcher` so it should include status objects in the watch stream if you have included these in the crd schema for `Sink`. A print thing like the above should also help confirm / deny this.

**Reply by sjbodzo** [reply]
The controller is spawned via `tokio::spawn` and has a `for_each` defined like: 
```
    ctr
        .shutdown_on_signal()
        .graceful_shutdown_on(shutdown_manager.wait_shutdown_triggered())
        .run(reconcile, error_policy, ctx).for_each(|result| async move {
            match result {
                Ok((app, _)) => tracing::debug!(%app, "processed event for app"),
                Err(err) => match err {
                 // exhaustive error match
```

The logs do not print when using a `predicate_filter`, making it clear that the events are somehow being ignored, as there are no logs indicating event receipt or processing. 

One other thing I failed to mention is that I have 3 different controllers on different custom resources running in the same code. I `spawn` each in my main. Could it be that some internal design using the `unstable` stream feature is thwarted by running multiple controllers? 

They are not sharing any components (watchers, etc) and the one with this odd masking issue is the only one trying to use the stream feature directly. 

Edit: Ive updated the logging a bit, and for `Controller::new` I clearly see the status write causing a _single_ re-trigger, presumably because the status written is idempotent and that subsequent attempts to write it are not an object update. I confirmed this assumption by editing away the entire status, and it only re-entered a single time to set it back. So that at least explains that aspect of things.

**Reply by clux** [reply]
Stream sharing could maybe play into this. What I expect would matter would be whether you are doing the filter before or after the line to call `reflect_shared` in the watcher stream chain. If it's possible to do `watcher().predicate_filter().reflect_shared()`* then the filter would in theory become global across all controllers (likely not what you want).

The stream sharing stuff is very new though, all the edge cases here are not experimented with (you may be the first to try predicates with shared streams).

**Reply by clux** [reply]
Actually, yeah. The stream sharing stuff has some unclear behaviour on how to broadcast events for cases like this. Basically, not everything is passed on atm as it was a bit unclear how broadcasts should interact with `touched_objects()` / `applied_objects()`. https://github.com/kube-rs/kube/blob/f9902f1439b3c0baafc2ece1680644c2bfade742/kube-runtime/src/reflector/store.rs#L138-L164 is current, and https://github.com/kube-rs/kube/issues/1472 is one follow-up ticket.

**Reply by sjbodzo** [reply]
Interesting. Thank you for all of the help and feedback. 

Im going to mark the above as the answer to highlight the important caveat, though in my case the watches are not shared between the controllers. The `for_stream` is using a stream isolated to its controller, separate from the other controllers (which both use standard `Controller::new` style).

The codebase in question is running on 0.88.1, which predates the introduction of the stream sharing functionality. I will play with this some more, and if anything interesting comes to light will share it here as follow-up. Thanks again!

**Reply by clux** [reply]
Ah, apologies. If it's only using the regular streams interface (and not shared streams), then your case is a little more suspicious. The general streams interface should not have as many rough edges.

You should be able to test this in isolation with just a watcher stream and a println after each element going through the predicate side without going through a controller (that side should not matter for you if the streams are not shared between controllers).

```rust
    let stream = watcher(api, watcher::Config::default().any_semantic())
        .default_backoff()
        .modify(|x| println!("pre-filter {}", x.name_any()))
        .applied_objects()
        .predicate_filter(predicates::resource_version);
    let mut stream = pin!(stream);

    while let Some(pod) = stream.try_next().await? {
        println!("post-filter {}", pod.name_any());
    }
```

**Reply by sjbodzo** [reply, marked as answer]
@ivan-kiselev identified our root issue. 

The first time a new object is seen, it does not have a finalizer set. Therefore [the code](https://github.com/kube-rs/kube/blob/main/kube-runtime/src/finalizer.rs#L201) will attempt to `patch` the object and `await_change`. Patching the object does not increment the object's `generation` so our predicate filter then prevents the object from triggering the controller. This is because the initial trigger caused generation = 1 to already be seen. 

Restarting the controller instance fixed the problem by presenting a new store to cache.  

The solution implemented was to write a custom predicate filter on `K` that inputs the hash of the `generation` (if present) _and_ then adds in the hash of the finalizers: 
```
pub fn generation_and_finalizers<K: Resource>(obj: &K) -> Option<u64> {
    let mut hasher = DefaultHasher::default();
    if let Some(gen) = obj.meta().generation {
        gen.hash(&mut hasher)
    };
    obj.finalizers().hash(&mut hasher);
    let hash = hasher.finish();
    Some(hash)
}
```

This prevents your stream-based controller from missing the initial object when using a `generation` predicate filter with a `finalizer` in your controller. 

@clux marking this as the solution for posterity

**Reply by clux** [reply]
Thanks for this! If it is of interest, we could add `predicates.finalizers` as a standalone fn, then you could create a predicate like your above fn in the future with `predicates::generation.fallback(predicates::finalizers)`.

**Reply by ivan-kiselev** [reply]
I will throw an item to my TODO list for this week to submit a PR, it doesn't seem to be much work :)

**Reply by ivan-kiselev** [reply]
here we go https://github.com/kube-rs/kube/pull/1560
